### ISSUE TITLE
[Feature] Multiple loggers

### ISSUE THREAD

**Issue by befelix** [meta-pytorch/forge#758]
### Context/Motivation

In our setup, we would like to track different metrics at different time-steps: While trainer metrics naturally fit into the `training_step`, performance scores (e.g., rewards on a validation dataset) would be ideally logged to the policy version that was being executed, rather than being arbitrarily chunked according to the `training_step`. However, such a setup is currently difficult given that the logger is a globally shared instance.

A practical example for why this would be useful: We observed drops in rewards after the first training step. This was the result of the replay buffer being empty initially and our generators completing shorter episodes first. This confusion would have been avoided if we been able to log the rewards to a specific policy version, instead of globally chunking all metrics at the `train_step` level.

**Comment by felipemello1** [comment]
If I understand correctly, the issue is about temporal misalignment between reward metrics and policy versions. 

```
generator is on policy version 0 (initial weights):
  - Rollout 1-A: short prompt → completes fast → reward = 0.8 → goes to buffer
  - Rollout 1-B: short prompt → completes fast → reward = 0.9 → goes to buffer
  train(), flush()
  - Rollout 1-C: long prompt → completes later → reward = 0.4 → goes to buffer
  - Rollout 1- D: long prompt → completes later → reward = 0.3 → goes to buffer
  train(), flush()
```

It looks like training made the policy worse (0.85 → 0.35), but both sets of rewards came from the same policy version 0.
The "drop" is just short-easy episodes finishing before long-hard ones.

----

When you say you want 2 loggers, do you mean you would like to generate 2 different wandb runs, but x-axis for 1 is train_step and x-axis for another is reward_step, or something like that?

I find it a bit complex. Would one of the ideas below suffice?

1. (RECOMMENDED) Log it with 'streaming mode', i.e. "logging_mode: per_rank_no_reduce". Its not **very** well tested, but its tested. 

This logs the metrics as soon as they are received, and then you can aggregate them as you prefer. You might want to change the backend to something other than wandb too.

docs: https://github.com/meta-pytorch/torchforge/tree/main/src/forge/observability#2-logging-modes 

2. (MITIGATION) See if there is something you can do in your code to attenuate the issue, e.g. accumulate rewards a,b,c,d and only them log the 'sum' and log some 'reward_step'. But those would still be aggregated per train_step.

You could see if there is a way to handle that with some custom accumulator: https://github.com/meta-pytorch/torchforge/blob/a111042db5a1a01d7cf464288e8958f8b0bdafb3/src/forge/observability/metrics.py#L421

3. (NOT RECOMMENDED) Change the position/frequency of flush 

i.e flush on train step, flush on reward step. Not sure how well this would work, as x-axis would be something odd now, but you can always manually change the x-axis in wandb to something else.

```
record_metric(myreward)
record_metric(reward_step) # manually in wandb do myreward x reward_step)
flush(step=1)

record_metric(train_loss)
record_metric(train_step) # manually in wandb do train_loss x train_step)
flush(step=2)
````
